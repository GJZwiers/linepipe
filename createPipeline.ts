import YAML from "npm:yaml";

export function validatePipelineName(name: string) {
  if (!/\.ya?ml$/.test(name)) {
    throw new Error(
      `Invalid name: ${name}. Pipeline must end with .yaml or .yml file extension`,
    );
  }
}

export function createPipeline(
  path: string,
  options: { canary: boolean },
  dir?: string,
) {
  const matrix = {
    os: ["ubuntu-22.04", "macos-12", "windows-2022"],
    "deno-version": ["v1.x"],
    include: (options.canary)
      ? [{ "deno-version": "canary", os: "ubuntu-22.04" }]
      : undefined,
  };

  const yamlObject = {
    name: "build",
    on: {
      push: {
        branches: ["main"],
      },
      pull_request: {
        branches: ["main"],
      },
    },
    jobs: {
      build: {
        "runs-on": "${{ matrix.os }}",
        "continue-on-error": (options.canary)
          ? '${{ matrix.deno-version == "canary" }}'
          : undefined,
        strategy: {
          matrix,
        },
        env: {
          DENO_DIR: "deno_dir",
        },
        steps: [
          {
            name: "Configure line-endings for Windows build",
            if: "matrix.os == 'windows-2022'",
            run:
              "git config --system core.autocrlf false\ngit config --system core.eol lf",
          },
          {
            uses: "actions/checkout@v3",
          },
          {
            uses: "denoland/setup-deno@v1.1.1",
            with: {
              "deno-version": "${{ matrix.deno-version }}",
            },
          },
          {
            run: "deno fmt --check",
          },
          {
            run: "deno lint",
          },
          {
            run: "deno check main.ts",
          },
          {
            name: "Cache DENO_DIR",
            uses: "actions/cache@v3",
            with: {
              path: "${{ env.DENO_DIR }}",
              key: "${{ hashFiles('deno.lock') }}",
            },
          },
          {
            run: "deno test --coverage=cov/",
          },
          {
            if:
              "matrix.os == 'ubuntu-22.04' && matrix.deno-version != 'canary'",
            run: "deno coverage --lcov --output=cov.lcov cov/",
          },
        ],
      },
    },
  };

  const yaml = YAML.stringify(yamlObject);
  const githubFolder = dir ? `${dir}/.github/workflows` : ".github/workflows";
  const data = new TextEncoder().encode(yaml);

  try {
    Deno.writeFileSync(`${githubFolder}/${path}`, data);
  } catch (e) {
    let writePath = path;
    if (e instanceof Deno.errors.NotFound) {
      writePath = `${githubFolder}/${path}`;
      Deno.mkdirSync(githubFolder, { recursive: true });
    } else if (e instanceof Deno.errors.PermissionDenied) {
      throw new Error(`Permission to ${githubFolder} was denied.`);
    }
    Deno.writeFileSync(writePath, data);
  }
}
