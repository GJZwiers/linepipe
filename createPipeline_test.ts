import { assertThrows } from "https://deno.land/std@0.177.0/testing/asserts.ts";
import {
  assertSnapshot,
} from "https://deno.land/std@0.177.0/testing/snapshot.ts";
import { createPipeline } from "./createPipeline.ts";

const test_dir = "test_dir";

Deno.test(async function throwsWhenNoGitHubDir() {
  await Deno.mkdir(test_dir);
  const path = `${test_dir}/build.yaml`;

  assertThrows(() => {
    createPipeline(path, { canary: false, githubDir: false });
  });

  await Deno.remove(test_dir, { recursive: true });
});

Deno.test(async function createsYAMLPipelineAndGitHubDir(context) {
  await Deno.mkdir(test_dir);
  const path = "build.yaml";

  const githubDir = ".github/workflows";
  createPipeline(path, { canary: false, githubDir: true }, test_dir);
  const file = await Deno.readFile(`${test_dir}/${githubDir}/${path}`);

  await assertSnapshot(context, new TextDecoder().decode(file));
  await Deno.remove(test_dir, { recursive: true });
});
