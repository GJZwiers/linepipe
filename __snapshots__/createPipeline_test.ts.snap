export const snapshot = {};

snapshot[`createsActionsPipeline 1`] = `
"name: build
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
jobs:
  build:
    runs-on: \${{ matrix.os }}
    strategy:
      matrix:
        os:
          - ubuntu-22.04
          - macos-12
          - windows-2022
        deno-version:
          - v1.x
    env:
      DENO_DIR: deno_dir
    steps:
      - name: Configure line-endings for Windows build
        if: matrix.os == 'windows-2022'
        run: |-
          git config --system core.autocrlf false
          git config --system core.eol lf
      - uses: actions/checkout@v3
      - uses: denoland/setup-deno@v1.1.1
        with:
          deno-version: \${{ matrix.deno-version }}
      - run: deno fmt --check
      - run: deno lint
      - run: deno check main.ts
      - name: Cache DENO_DIR
        uses: actions/cache@v3
        with:
          path: \${{ env.DENO_DIR }}
          key: \${{ hashFiles('deno.lock') }}
      - run: deno test --coverage=cov/
      - if: matrix.os == 'ubuntu-22.04' && matrix.deno-version != 'canary'
        run: deno coverage --lcov --output=cov.lcov cov/
"
`;
